File: preprocessing/evaluation/evaluator.py
import json
import sys
import os
from collections import defaultdict

# Add parent directory to path
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from search_algorithms.tfidf_search import TFIDFSearchEngine
from search_algorithms.bm25_search import BM25SearchEngine


class SearchEvaluator:
    """Evaluasi performa algoritma pencarian"""
    
    def __init__(self, tfidf_engine, bm25_engine):
        self.tfidf_engine = tfidf_engine
        self.bm25_engine = bm25_engine
    
    def calculate_precision(self, retrieved_docs, relevant_docs):
        """
        Precision = (Retrieved AND Relevant) / Retrieved
        """
        if len(retrieved_docs) == 0:
            return 0.0
        
        relevant_retrieved = len(set(retrieved_docs) & set(relevant_docs))
        return relevant_retrieved / len(retrieved_docs)
    
    def calculate_recall(self, retrieved_docs, relevant_docs):
        """
        Recall = (Retrieved AND Relevant) / Relevant
        """
        if len(relevant_docs) == 0:
            return 0.0
        
        relevant_retrieved = len(set(retrieved_docs) & set(relevant_docs))
        return relevant_retrieved / len(relevant_docs)
    
    def calculate_f1_score(self, precision, recall):
        """
        F1 Score = 2 * (Precision * Recall) / (Precision + Recall)
        """
        if precision + recall == 0:
            return 0.0
        
        return 2 * (precision * recall) / (precision + recall)
    
    def calculate_average_precision(self, retrieved_docs, relevant_docs):
        """
        Average Precision (AP) untuk satu query
        AP = (1/R) * Σ (Precision@k * rel(k))
        """
        if len(relevant_docs) == 0:
            return 0.0
        
        relevant_set = set(relevant_docs)
        precision_sum = 0.0
        relevant_count = 0
        
        for k, doc_id in enumerate(retrieved_docs, 1):
            if doc_id in relevant_set:
                relevant_count += 1
                precision_at_k = relevant_count / k
                precision_sum += precision_at_k
        
        return precision_sum / len(relevant_docs) if len(relevant_docs) > 0 else 0.0
    
    def calculate_map(self, query_results, ground_truth):
        """
        Mean Average Precision (MAP) across all queries
        MAP = (1/Q) * Σ AP(q)
        """
        if len(query_results) == 0:
            return 0.0
        
        ap_sum = 0.0
        for query, retrieved_docs in query_results.items():
            relevant_docs = ground_truth.get(query, [])
            ap = self.calculate_average_precision(retrieved_docs, relevant_docs)
            ap_sum += ap
        
        return ap_sum / len(query_results)
    
    def evaluate_single_query(self, query, relevant_docs, top_k=10):
        """Evaluasi untuk satu query dengan kedua algoritma"""
        # TF-IDF results
        tfidf_results = self.tfidf_engine.search(query, top_k=top_k)
        tfidf_doc_ids = [r['doc_id'] for r in tfidf_results]
        
        # BM25 results
        bm25_results = self.bm25_engine.search(query, top_k=top_k)
        bm25_doc_ids = [r['doc_id'] for r in bm25_results]
        
        # Calculate metrics for TF-IDF
        tfidf_precision = self.calculate_precision(tfidf_doc_ids, relevant_docs)
        tfidf_recall = self.calculate_recall(tfidf_doc_ids, relevant_docs)
        tfidf_f1 = self.calculate_f1_score(tfidf_precision, tfidf_recall)
        tfidf_ap = self.calculate_average_precision(tfidf_doc_ids, relevant_docs)
        
        # Calculate metrics for BM25
        bm25_precision = self.calculate_precision(bm25_doc_ids, relevant_docs)
        bm25_recall = self.calculate_recall(bm25_doc_ids, relevant_docs)
        bm25_f1 = self.calculate_f1_score(bm25_precision, bm25_recall)
        bm25_ap = self.calculate_average_precision(bm25_doc_ids, relevant_docs)
        
        return {
            'query': query,
            'tfidf': {
                'precision': tfidf_precision,
                'recall': tfidf_recall,
                'f1_score': tfidf_f1,
                'average_precision': tfidf_ap,
                'results': tfidf_results
            },
            'bm25': {
                'precision': bm25_precision,
                'recall': bm25_recall,
                'f1_score': bm25_f1,
                'average_precision': bm25_ap,
                'results': bm25_results
            }
        }
    
    def evaluate_multiple_queries(self, test_queries, top_k=10):
        """
        Evaluasi untuk multiple queries
        
        test_queries format: {
            "query1": ["doc_id1", "doc_id2", ...],  # relevant docs
            "query2": ["doc_id3", "doc_id4", ...],
            ...
        }
        """
        results = []
        
        tfidf_all_results = {}
        bm25_all_results = {}
        
        for query, relevant_docs in test_queries.items():
            evaluation = self.evaluate_single_query(query, relevant_docs, top_k)
            results.append(evaluation)
            
            # Collect for MAP calculation
            tfidf_all_results[query] = [r['doc_id'] for r in evaluation['tfidf']['results']]
            bm25_all_results[query] = [r['doc_id'] for r in evaluation['bm25']['results']]
        
        # Calculate MAP
        tfidf_map = self.calculate_map(tfidf_all_results, test_queries)
        bm25_map = self.calculate_map(bm25_all_results, test_queries)
        
        # Calculate average metrics
        avg_metrics = {
            'tfidf': {
                'avg_precision': sum(r['tfidf']['precision'] for r in results) / len(results),
                'avg_recall': sum(r['tfidf']['recall'] for r in results) / len(results),
                'avg_f1_score': sum(r['tfidf']['f1_score'] for r in results) / len(results),
                'map': tfidf_map
            },
            'bm25': {
                'avg_precision': sum(r['bm25']['precision'] for r in results) / len(results),
                'avg_recall': sum(r['bm25']['recall'] for r in results) / len(results),
                'avg_f1_score': sum(r['bm25']['f1_score'] for r in results) / len(results),
                'map': bm25_map
            }
        }
        
        return {
            'individual_results': results,
            'average_metrics': avg_metrics
        }
    
    def print_comparison_report(self, evaluation_results):
        """Print formatted comparison report"""
        avg_metrics = evaluation_results['average_metrics']
        
        print("\n" + "="*80)
        print("SEARCH ALGORITHM COMPARISON REPORT")
        print("="*80)
        
        print("\n" + "-"*80)
        print("AVERAGE METRICS ACROSS ALL QUERIES")
        print("-"*80)
        
        print(f"\n{'Metric':<25} {'TF-IDF':>15} {'BM25':>15} {'Winner':>15}")
        print("-"*80)
        
        # Precision
        tfidf_prec = avg_metrics['tfidf']['avg_precision']
        bm25_prec = avg_metrics['bm25']['avg_precision']
        winner_prec = "TF-IDF" if tfidf_prec > bm25_prec else "BM25" if bm25_prec > tfidf_prec else "Tie"
        print(f"{'Precision':<25} {tfidf_prec:>15.4f} {bm25_prec:>15.4f} {winner_prec:>15}")
        
        # Recall
        tfidf_rec = avg_metrics['tfidf']['avg_recall']
        bm25_rec = avg_metrics['bm25']['avg_recall']
        winner_rec = "TF-IDF" if tfidf_rec > bm25_rec else "BM25" if bm25_rec > tfidf_rec else "Tie"
        print(f"{'Recall':<25} {tfidf_rec:>15.4f} {bm25_rec:>15.4f} {winner_rec:>15}")
        
        # F1 Score
        tfidf_f1 = avg_metrics['tfidf']['avg_f1_score']
        bm25_f1 = avg_metrics['bm25']['avg_f1_score']
        winner_f1 = "TF-IDF" if tfidf_f1 > bm25_f1 else "BM25" if bm25_f1 > tfidf_f1 else "Tie"
        print(f"{'F1 Score':<25} {tfidf_f1:>15.4f} {bm25_f1:>15.4f} {winner_f1:>15}")
        
        # MAP
        tfidf_map = avg_metrics['tfidf']['map']
        bm25_map = avg_metrics['bm25']['map']
        winner_map = "TF-IDF" if tfidf_map > bm25_map else "BM25" if bm25_map > tfidf_map else "Tie"
        print(f"{'MAP':<25} {tfidf_map:>15.4f} {bm25_map:>15.4f} {winner_map:>15}")
        
        print("\n" + "="*80)
        print("INDIVIDUAL QUERY RESULTS")
        print("="*80)
        
        for i, result in enumerate(evaluation_results['individual_results'], 1):
            print(f"\nQuery {i}: '{result['query']}'")
            print("-"*80)
            print(f"{'Metric':<25} {'TF-IDF':>15} {'BM25':>15}")
            print("-"*80)
            print(f"{'Precision':<25} {result['tfidf']['precision']:>15.4f} {result['bm25']['precision']:>15.4f}")
            print(f"{'Recall':<25} {result['tfidf']['recall']:>15.4f} {result['bm25']['recall']:>15.4f}")
            print(f"{'F1 Score':<25} {result['tfidf']['f1_score']:>15.4f} {result['bm25']['f1_score']:>15.4f}")
            print(f"{'Average Precision':<25} {result['tfidf']['average_precision']:>15.4f} {result['bm25']['average_precision']:>15.4f}")


if __name__ == "__main__":
    # Test evaluation
    index_path = '../dataset/inverted_index.json'
    corpus_path = '../dataset/stemmed_corpus.json'
    
    print("Loading search engines...")
    tfidf_engine = TFIDFSearchEngine(index_path, corpus_path)
    bm25_engine = BM25SearchEngine(index_path, corpus_path)
    
    evaluator = SearchEvaluator(tfidf_engine, bm25_engine)
    
    # Example test queries dengan ground truth
    # NOTE: Ini contoh, Anda perlu menyesuaikan dengan doc_id yang sebenarnya
    test_queries = {
        "perang di Myanmar": [],  # Isi dengan doc_id yang relevan
        "konflik etnis Rohingya": [],
        "junta militer kudeta": [],
    }
    
    print("\nNote: Please update test_queries with actual relevant doc_ids for proper evaluation.")
    print("Example format:")
    print("test_queries = {")
    print("    'query1': ['doc_id_1', 'doc_id_2', ...],")
    print("    'query2': ['doc_id_3', 'doc_id_4', ...],")
    print("}")

File: lib/evaluation/evaluator.ts
import { tfidfSearchAsync } from '../search/tfidf-search';
import { bm25SearchAsync } from '../search/bm25-search';
import { groundTruthData, QueryGroundTruth } from './ground-truth';

/**
 * Hasil evaluasi untuk satu query
 */
export interface QueryEvaluationResult {
  query: string;
  precision: number;
  recall: number;
  f1Score: number;
  averagePrecision: number;
  retrievedDocs: number[];
  relevantDocs: number[];
  relevantRetrieved: number[];
}

/**
 * Hasil evaluasi keseluruhan untuk satu algoritma
 */
export interface AlgorithmEvaluationResult {
  algorithm: 'TF-IDF' | 'BM25';
  precision: number;
  recall: number;
  f1Score: number;
  map: number; // Mean Average Precision
  queryResults: QueryEvaluationResult[];
}

/**
 * Hasil evaluasi komparasi kedua algoritma
 */
export interface EvaluationComparison {
  tfidf: AlgorithmEvaluationResult;
  bm25: AlgorithmEvaluationResult;
  timestamp: string;
  totalQueries: number;
}

/**
 * Menghitung Precision: TP / (TP + FP)
 * Proporsi dokumen yang diambil yang benar-benar relevan
 */
function calculatePrecision(retrieved: number[], relevant: number[]): number {
  if (retrieved.length === 0) return 0;
  
  const relevantRetrieved = retrieved.filter(doc => relevant.includes(doc));
  return relevantRetrieved.length / retrieved.length;
}

/**
 * Menghitung Recall: TP / (TP + FN)
 * Proporsi dokumen relevan yang berhasil diambil
 */
function calculateRecall(retrieved: number[], relevant: number[]): number {
  if (relevant.length === 0) return 0;
  
  const relevantRetrieved = retrieved.filter(doc => relevant.includes(doc));
  return relevantRetrieved.length / relevant.length;
}

/**
 * Menghitung F1 Score: 2 * (Precision * Recall) / (Precision + Recall)
 * Harmonic mean dari Precision dan Recall
 */
function calculateF1Score(precision: number, recall: number): number {
  if (precision + recall === 0) return 0;
  return (2 * precision * recall) / (precision + recall);
}

/**
 * Menghitung Average Precision untuk satu query
 * AP = (1/R) * Σ(P(k) * rel(k))
 * dimana R = jumlah dokumen relevan, P(k) = precision at k, rel(k) = 1 jika relevan
 */
function calculateAveragePrecision(retrieved: number[], relevant: number[]): number {
  if (relevant.length === 0) return 0;
  
  let sumPrecision = 0;
  let relevantCount = 0;
  
  for (let i = 0; i < retrieved.length; i++) {
    const doc = retrieved[i];
    if (relevant.includes(doc)) {
      relevantCount++;
      const precisionAtK = relevantCount / (i + 1);
      sumPrecision += precisionAtK;
    }
  }
  
  return sumPrecision / relevant.length;
}

/**
 * Evaluasi satu query dengan algoritma tertentu
 */
async function evaluateQuery(
  groundTruth: QueryGroundTruth,
  searchFunction: (query: string, topK: number) => Promise<any[]>,
  topK: number = 20
): Promise<QueryEvaluationResult> {
  try {
    // Ambil hasil pencarian
    const searchResults = await searchFunction(groundTruth.query, topK);
    
    // Extract doc_id from search results (handle both id and doc_id properties)
    const retrievedDocs = searchResults.map(result => {
      // Parse doc_id to number if it's a string like "123"
      const docIdStr = result.doc_id || result.id || '';
      const docIdNum = parseInt(docIdStr.toString().replace(/\D/g, ''), 10);
      return isNaN(docIdNum) ? 0 : docIdNum;
    }).filter(id => id > 0);
    
    const relevantDocs = groundTruth.relevantDocs;
    
    // Hitung dokumen relevan yang berhasil diambil
    const relevantRetrieved = retrievedDocs.filter(doc => relevantDocs.includes(doc));
    
    // Hitung metrics
    const precision = calculatePrecision(retrievedDocs, relevantDocs);
    const recall = calculateRecall(retrievedDocs, relevantDocs);
    const f1Score = calculateF1Score(precision, recall);
    const averagePrecision = calculateAveragePrecision(retrievedDocs, relevantDocs);
    
    return {
      query: groundTruth.query,
      precision,
      recall,
      f1Score,
      averagePrecision,
      retrievedDocs,
      relevantDocs,
      relevantRetrieved
    };
  } catch (error) {
    console.error(`Error evaluating query "${groundTruth.query}":`, error);
    return {
      query: groundTruth.query,
      precision: 0,
      recall: 0,
      f1Score: 0,
      averagePrecision: 0,
      retrievedDocs: [],
      relevantDocs: groundTruth.relevantDocs,
      relevantRetrieved: []
    };
  }
}

/**
 * Evaluasi semua query untuk satu algoritma
 */
async function evaluateAlgorithm(
  algorithm: 'TF-IDF' | 'BM25',
  searchFunction: (query: string, topK: number) => Promise<any[]>,
  topK: number = 20
): Promise<AlgorithmEvaluationResult> {
  const queryResults: QueryEvaluationResult[] = [];
  
  // Evaluasi setiap query
  for (const groundTruth of groundTruthData) {
    const result = await evaluateQuery(groundTruth, searchFunction, topK);
    queryResults.push(result);
  }
  
  // Hitung rata-rata metrics
  const avgPrecision = queryResults.reduce((sum, r) => sum + r.precision, 0) / queryResults.length;
  const avgRecall = queryResults.reduce((sum, r) => sum + r.recall, 0) / queryResults.length;
  const avgF1Score = queryResults.reduce((sum, r) => sum + r.f1Score, 0) / queryResults.length;
  const map = queryResults.reduce((sum, r) => sum + r.averagePrecision, 0) / queryResults.length;
  
  return {
    algorithm,
    precision: avgPrecision,
    recall: avgRecall,
    f1Score: avgF1Score,
    map,
    queryResults
  };
}

/**
 * Evaluasi dan bandingkan kedua algoritma
 */
export async function evaluateSearchEngines(topK: number = 20): Promise<EvaluationComparison> {
  console.log('Starting evaluation of search engines...');
  
  // Evaluasi TF-IDF
  console.log('Evaluating TF-IDF...');
  const tfidfResults = await evaluateAlgorithm('TF-IDF', tfidfSearchAsync, topK);
  
  // Evaluasi BM25
  console.log('Evaluating BM25...');
  const bm25Results = await evaluateAlgorithm('BM25', bm25SearchAsync, topK);
  
  console.log('Evaluation completed!');
  
  return {
    tfidf: tfidfResults,
    bm25: bm25Results,
    timestamp: new Date().toISOString(),
    totalQueries: groundTruthData.length
  };
}

/**
 * Format hasil evaluasi untuk ditampilkan
 */
export function formatEvaluationResults(comparison: EvaluationComparison): string {
  let output = '=== EVALUATION RESULTS ===\n\n';
  output += `Total Queries: ${comparison.totalQueries}\n`;
  output += `Timestamp: ${comparison.timestamp}\n\n`;
  
  // TF-IDF Results
  output += '--- TF-IDF Performance ---\n';
  output += `Precision: ${(comparison.tfidf.precision * 100).toFixed(2)}%\n`;
  output += `Recall: ${(comparison.tfidf.recall * 100).toFixed(2)}%\n`;
  output += `F1 Score: ${(comparison.tfidf.f1Score * 100).toFixed(2)}%\n`;
  output += `MAP: ${(comparison.tfidf.map * 100).toFixed(2)}%\n\n`;
  
  // BM25 Results
  output += '--- BM25 Performance ---\n';
  output += `Precision: ${(comparison.bm25.precision * 100).toFixed(2)}%\n`;
  output += `Recall: ${(comparison.bm25.recall * 100).toFixed(2)}%\n`;
  output += `F1 Score: ${(comparison.bm25.f1Score * 100).toFixed(2)}%\n`;
  output += `MAP: ${(comparison.bm25.map * 100).toFixed(2)}%\n\n`;
  
  // Comparison
  output += '--- Comparison ---\n';
  const precisionDiff = (comparison.bm25.precision - comparison.tfidf.precision) * 100;
  const recallDiff = (comparison.bm25.recall - comparison.tfidf.recall) * 100;
  const f1Diff = (comparison.bm25.f1Score - comparison.tfidf.f1Score) * 100;
  const mapDiff = (comparison.bm25.map - comparison.tfidf.map) * 100;
  
  output += `Precision: BM25 ${precisionDiff > 0 ? '+' : ''}${precisionDiff.toFixed(2)}% vs TF-IDF\n`;
  output += `Recall: BM25 ${recallDiff > 0 ? '+' : ''}${recallDiff.toFixed(2)}% vs TF-IDF\n`;
  output += `F1 Score: BM25 ${f1Diff > 0 ? '+' : ''}${f1Diff.toFixed(2)}% vs TF-IDF\n`;
  output += `MAP: BM25 ${mapDiff > 0 ? '+' : ''}${mapDiff.toFixed(2)}% vs TF-IDF\n`;
  
  return output;
}
